### PR TITLE
Add additional fields for ServiceRole

### PR DESCRIPTION
**Pull request by pitlv2109**
Support notPaths, notMethods, ports, notPorts, hosts, notHosts in ServiceRole (+ unit testing and refactoring).

PR for #11516.
Thank you Yangmin for the help.

### CODE REVIEW

**Comment by pitlv2109** [comment]
_prow/istio-pilot-multicluster-e2e.sh_ failed similar to merged PR https://github.com/istio/istio/pull/11704
_prow/istio-integ-k8s-tests.sh_ failed similar to merged PR https://github.com/istio/istio/pull/11627
_prow/release-test.sh_ said no test failures. Seems like something's wrong with the built when the test ran? 
All pilot unit tests passed locally.

**Comment by yangminzhu** [comment]
We'll need to fix the `ci/circleci: test` and `prow/release-test.sh`, could you take a look if it's related to your PR? If not, we may need to sync with release-1.1 again to pick up any fixes.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 1]
```diff
@@ -557,6 +560,26 @@ func convertRbacRulesToFilterConfig(service *serviceMetadata, option rbacOption)
...
+// appendRule appends a |rule| to |rules| if |rule| is not nil.
+func appendRule(rules *policyproto.Permission_AndRules, rule *policyproto.Permission) {
+	if rule == nil {
+		return
+	}
+	rules.AndRules.Rules = append(rules.AndRules.Rules, rule)
+}
+
+// appendNotRule appends a |notRule| to |rules| if |notRule| is not nil.
+func appendNotRule(rules *policyproto.Permission_AndRules, notRule *policyproto.Permission) {
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: The comment should be more clear that it wraps the `notRule` inside a `AndRule` and then append to `rules`.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 2]
```diff
@@ -565,28 +588,52 @@ func convertToPermission(rule *rbacproto.AccessRule) *policyproto.Permission {
...
 
 	if len(rule.Constraints) > 0 {
 		// Constraint rule is matched with AND semantics, it's invalid if 2 constraints have the same
 		// key and this should already be caught in validation stage.
 		for _, constraint := range rule.Constraints {
 			p := permissionForKeyValues(constraint.Key, constraint.Values)
-			if p != nil {
-				rules.AndRules.Rules = append(rules.AndRules.Rules, p)
-			}
+			appendRule(rules, p)
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: Can we also use `appendRule` in line 642?

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 3]
```diff
@@ -430,6 +430,61 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+				},
+			},
+		},
+		{
+			ConfigMeta: model.ConfigMeta{Name: "service-role-10"},
+			Spec: &rbacproto.ServiceRole{
+				Rules: []*rbacproto.AccessRule{
+					{
+						Services: []string{"service-10"},
+						Hosts:    []string{"104.198.191.125"},
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: nit: maybe use a real host string.

File: pilot/pkg/networking/plugin/authz/rbac_test.go [hunk 4]
```diff
@@ -430,6 +430,61 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
...
+				},
+			},
+		},
+		{
+			ConfigMeta: model.ConfigMeta{Name: "service-role-11"},
+			Spec: &rbacproto.ServiceRole{
+				Rules: []*rbacproto.AccessRule{
+					{
+						Services: []string{"service-11"},
+						NotHosts: []string{"104.198.191.125"},
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac_test.go]: nit: same as above.

File: pilot/pkg/networking/plugin/authz/test_helper.go [hunk 5]
```diff
@@ -189,3 +190,73 @@ func generatePolicyWithHTTPMethodAndGroupClaim(methodName, claimName string) *po
 		}},
 	}
 }
+
+// nolint:deadcode
+func generateExpectRBACForSinglePolicy(serviceRoleID string, rbacPolicy *policy.Policy) *policy.RBAC {
+	return &policy.RBAC{
+		Action: policy.RBAC_ALLOW,
+		Policies: map[string]*policy.Policy{
+			"service-role-" + serviceRoleID: rbacPolicy,
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/test_helper.go]: I suggest we don't hard-code the name here, let the caller to put the full name when calling. This makes the code more clear and self-explanation. And it handles if we want to use a different name.

File: pilot/pkg/networking/plugin/authz/test_helper.go [hunk 6]
```diff
@@ -189,3 +190,73 @@ func generatePolicyWithHTTPMethodAndGroupClaim(methodName, claimName string) *po
...
+			Subjects: []*rbacproto.Subject{
+				{
+					Properties: map[string]string{
+						"request.auth.claims[groups]": "group*",
+					},
+				},
+			},
+			RoleRef: &rbacproto.RoleRef{
+				Kind: "ServiceRole",
+				Name: "service-role-" + serviceRoleOrBindingID,
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/test_helper.go]: Same as above.

File: pilot/pkg/networking/plugin/authz/test_helper.go [hunk 7]
```diff
@@ -189,3 +190,73 @@ func generatePolicyWithHTTPMethodAndGroupClaim(methodName, claimName string) *po
...
+			"service-role-" + serviceRoleID: rbacPolicy,
+		},
+	}
+}
+
+// nolint:deadcode
+// serviceRoleOrBindingID: A number represent ServiceRole or ServiceRoleBinding number.
+func generateSimpleServiceRoleBindingAllGroups(serviceRoleOrBindingID string) model.Config {
+	return model.Config{
+		ConfigMeta: model.ConfigMeta{Name: "service-role-binding-" + serviceRoleOrBindingID},
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/test_helper.go]: Same as above.

File: pilot/pkg/networking/plugin/authz/util.go [hunk 8]
```diff
@@ -102,6 +102,14 @@ func convertToPort(v string) (uint32, error) {
 	return uint32(p), nil
 }
 
+func convertPortInIntToString(intPorts []int32) []string {
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/util.go]: `convertPortInIntToString` -> `convertPortsToString`

File: pilot/pkg/networking/plugin/authz/util_test.go [hunk 9]
```diff
@@ -170,6 +170,35 @@ func TestConvertToPort(t *testing.T) {
...
+		{
+			Name:   "valid ports",
+			V:      []int32{80, 3000, 443},
+			Expect: []string{"80", "3000", "443"},
+		},
+		{
+			Name:   "valid port",
+			V:      []int32{9080},
+			Expect: []string{"9080"},
+		},
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/util_test.go]: Also add some invalid cases.

File: pilot/pkg/model/validation.go [hunk 10]
```diff
@@ -1460,6 +1471,16 @@ func ValidateServiceRole(name, namespace string, msg proto.Message) error {
 	return errs
 }
 
+// arePortsInRange checks if all ports in range [0, 65535]
+func arePortsInRange(ports []int32) bool {
```
Comment (yangminzhu) [inline pilot/pkg/model/validation.go]: You can just reuse `ValidatePort()` in this file.

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 11]
```diff
@@ -80,8 +80,11 @@ const (
 	attrDestUser      = "destination.user"      // service account, e.g. "bookinfo-productpage".
 	attrConnSNI       = "connection.sni"        // server name indication, e.g. "www.example.com".
 
+	// attributes that could be used in a ServiceRole rules.
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authz/rbac.go]: What does this comment mean? User cannot use these pseudo attributes in ServiceRole. It's actually internal implementation detail about how do we generate them in Envoy config.

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 1 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/rbac.go]: Done!

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 2 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/rbac.go]: Done!

File: pilot/pkg/networking/plugin/authz/rbac_test.go (hunk 3 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: Done!

File: pilot/pkg/networking/plugin/authz/rbac_test.go (hunk 4 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/rbac_test.go]: Done!

File: pilot/pkg/networking/plugin/authz/test_helper.go (hunk 5 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/test_helper.go]: Done!

File: pilot/pkg/networking/plugin/authz/test_helper.go (hunk 6 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/test_helper.go]: Done!

File: pilot/pkg/networking/plugin/authz/test_helper.go (hunk 7 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/test_helper.go]: Done!

File: pilot/pkg/networking/plugin/authz/util.go (hunk 8 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/util.go]: Done!

File: pilot/pkg/model/validation.go (hunk 10 above)
Comment (pitlv2109) [reply pilot/pkg/model/validation.go]: Ah, I didn't see this function. Will do. Thanks!

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 11 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/rbac.go]: Sorry it was confusing. Edited!

File: pilot/pkg/networking/plugin/authz/util_test.go (hunk 9 above)
Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/util_test.go]: This is simply converting ints to strings. strconv.Itoa will always succeed if the input type is valid. Are you worried about the range of the ports? If so, we have already checked it before this function is called, right?

Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/util_test.go]: I see, then we don't need it here

Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/util_test.go]: Thank you Yangmin

File: pilot/pkg/networking/plugin/authz/test_helper.go [hunk 12]
```diff
@@ -189,3 +190,72 @@ func generatePolicyWithHTTPMethodAndGroupClaim(methodName, claimName string) *po
...
+			},
+		},
+		Principals: []*policy.Principal{{
+			Identifier: &policy.Principal_AndIds{
+				AndIds: &policy.Principal_Set{
+					Ids: []*policy.Principal{
+						{
+							Identifier: &policy.Principal_Metadata{
+								Metadata: generateMetadataListMatcher(authn.AuthnFilterName,
+									[]string{attrRequestClaims, "groups"}, "group*"),
```
Comment (liminw) [inline pilot/pkg/networking/plugin/authz/test_helper.go]: NIT: I saw "group*" has been used in the tests in many places. We probably do not need to test the same thing in so many tests. For example, if a test is for "not_methods" and not for testing "principal", let's use the simplest principal (e.g., a single service account).

Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/test_helper.go]: Thanks Limin. Maybe another PR? I still need to submit another small PR to check if the user defines not_ports and destination.port at the same time.

Comment (liminw) [reply pilot/pkg/networking/plugin/authz/test_helper.go]: Sure. Just a small suggestion. You don't have to make the change now.

Comment (pitlv2109) [reply pilot/pkg/networking/plugin/authz/test_helper.go]: Thanks!

**Pull request merged**